src/app/tests/TestAttributeCache.cpp

```diff
@@ -612,10 +612,45 @@ void TestCache(nlTestSuite * apSuite, void * apContext)
                              AttributeInstruction(AttributeInstruction::kAttributeB, 0, AttributeInstruction::kData) });
 }
 
+void TestSortDataVersionFilterSet(nlTestSuite * apSuite, void * apContext)
```

Comment:
These tests are in-sufficient. We need representative, pseudo end-to-end tests for a feature like this, and that can only be made possible by adding tests to `src/controller/data_model`.

Reply:
I have added several E2E tests in TestReadInteraction.cpp, and would add E2E test in controller/data_model as well, thanks

Reply:
The specific tests I'd like to see:

1. A wildcard read to an entire endpoint. On the server, there are 3 clusters. The very first read with an empty cache should not contain any items in the data filter list as received by the server. On a subsequent read attempt, we should see 3 items in the data filter list. Consequently, we should see no calls to `OnAttributeData` called on the client.
3. A read to an entire endpoint. On the server, there are a 100 cluster instances. The very first read with an empty cache should not contain any items in the data filter list. On a subsequent read attempt, it should maximally pack as many paths as it can before it  gracefully does rollback. We should structure it such that each cluster instance from 0 to 100 has slightly bigger bits of attribute data, so that we can then trivially validate that the paths that were sent were the X largest clusters.

Reply:
done, thanks